File: python/aoc23/day05/shared.py
from collections import defaultdict
import typing


def _parse_input(
    input_data: typing.TextIO,
) -> tuple[list[int], list[list[tuple[int, int, int]]]]:
    seeds: list[int] | None = None
    mapData: defaultdict[str, list[tuple[int, int, int]]] = defaultdict(list)
    for line in input_data:
        line = line.strip()
        if not line:
            continue
        elif line.startswith("seeds"):
            seeds = list(map(int, line.split()[1:]))
            break
    currentGroup: list[tuple[int, int, int]] | None = None
    for line in input_data:
        line = line.strip()
        if not line:
            continue
        line_data: list[str] = line.split()
        match line_data[0]:
            case "seed-to-soil":
                currentGroup = mapData[line_data[0]]
            case "soil-to-fertilizer":
                currentGroup = mapData[line_data[0]]
            case "fertilizer-to-water":
                currentGroup = mapData[line_data[0]]
            case "water-to-light":
                currentGroup = mapData[line_data[0]]
            case "light-to-temperature":
                currentGroup = mapData[line_data[0]]
            case "temperature-to-humidity":
                currentGroup = mapData[line_data[0]]
            case "humidity-to-location":
                currentGroup = mapData[line_data[0]]
            case _:
                values = tuple(map(int, line_data))
                if len(values) != 3:
                    raise ValueError(f"Expected 3 integers in line: {line}")
                if currentGroup is None:
                    raise ValueError(f"Data line before any key: {line}")
                currentGroup.append(values)

    key_sequence = [
        "seed-to-soil",
        "soil-to-fertilizer",
        "fertilizer-to-water",
        "water-to-light",
        "light-to-temperature",
        "temperature-to-humidity",
        "humidity-to-location",
    ]
    transform_sequence: list[list[tuple[int, int, int]]] = []
    for key in key_sequence:
        if key not in mapData:
            raise ValueError(f"Missing key in input: {key}")
        transform_sequence.append(mapData[key])

    return seeds or [], transform_sequence

File: cpp/src/aoc23/day05/shared.cpp
#include "day05.h"

namespace aoc23::day05
{
    int parseInput(std::istream& in, aoc23::day05::InputData& data)
    {
        std::string line;
        std::size_t lastPos, pos;
        std::size_t currentMapIdx = 0;
        while (std::getline(in, line))
        {
            if (line.empty())
            {
                continue;
            }
            size_t rowIdx = line.find(":");
            if (rowIdx != std::string::npos)
            {
                if (line.starts_with("seed-to-soil"))
                {
                    currentMapIdx = 0;
                    continue;
                }
                if (line.starts_with("soil-to-fertilizer"))
                {
                    currentMapIdx = 1;
                    continue;
                }
                if (line.starts_with("fertilizer-to-water"))
                {
                    currentMapIdx = 2;
                    continue;
                }
                if (line.starts_with("water-to-light"))
                {
                    currentMapIdx = 3;
                    continue;
                }
                if (line.starts_with("light-to-temperature"))
                {
                    currentMapIdx = 4;
                    continue;
                }
                if (line.starts_with("temperature-to-humidity"))
                {
                    currentMapIdx = 5;
                    continue;
                }
                if (line.starts_with("humidity-to-location"))
                {
                    currentMapIdx = 6;
                    continue;
                }
                if (line.starts_with("seeds"))
                {
                    currentMapIdx = 7;
                    std::size_t cPos = rowIdx + 1;
                    while (cPos < line.length())
                    {
                        while (cPos < line.length() && line[cPos] == ' ')
                            cPos++;
                        if (cPos >= line.length())
                            break;
                        std::size_t endPos = line.find(' ', cPos);
                        if (endPos == std::string::npos)
                            endPos = line.length();
                        long long number = std::stoll(line.substr(cPos, endPos - cPos));
                        data.seeds.push_back(number);
                        cPos = endPos;
                    }
                    continue;
                }
            }
            // potentially line
            if (currentMapIdx == 7)
            {
                continue;
            }

            size_t cPos = 0;
            long long parsedNumbers[3] = {0, 0, 0};
            int currentNumberIdx = 0;
            while (cPos < line.length() && currentNumberIdx < 3)
            {
                while (cPos < line.length() && line[cPos] == ' ')
                    cPos++;
                if (cPos >= line.length())
                    break;
                std::size_t endPos = line.find(' ', cPos);
                if (endPos == std::string::npos)
                    endPos = line.length();
                parsedNumbers[currentNumberIdx] = std::stoll(line.substr(cPos, endPos - cPos));
                cPos = endPos;
                currentNumberIdx++;
            }
            if (currentNumberIdx == 3)
            {
                std::array<long long, 3> arr = {parsedNumbers[0], parsedNumbers[1],
                                                parsedNumbers[2]};
                data.mapData[currentMapIdx].push_back(arr);
            }
        }
        return 0;
    }
}  // namespace aoc23::day05